analyzers/src/SonarAnalyzer.CSharp/Rules/UnusedReturnValue.cs

```diff
@@ -35,105 +35,89 @@ namespace SonarAnalyzer.Rules.CSharp
-                    {
-                        return;
-                    }
-
-                    var matchingInvocations = GetLocalMatchingInvocations(topmostContainingMethod, localFunctionSymbol, c.SemanticModel);
-                    if (IsNoncompliant(matchingInvocations))
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, localFunctionSyntax.ReturnType.GetLocation()));
-                    }
-                },
-                SyntaxKindEx.LocalFunctionStatement);
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, declaredPrivateMethodWithReturn.SyntaxNode.ReturnType.GetLocation()));
+                }
+            }
         }
 
-        /// <summary>
-        /// Method invocation is noncompliant when there is at least 1 inovcation of the method, and no invocation is
-        /// using the return value. The case of 0 invocation is handled by S1144 <see cref="UnusedPrivateMember"/>
-        /// </summary>
-        private static bool IsNoncompliant(IEnumerable<SyntaxNodeSymbolSemanticModelTuple<InvocationExpressionSyntax, IMethodSymbol>> matchingInvocations)
+        private static void AnalyzeLocalFunctionStatements(SyntaxNodeAnalysisContext context)
         {
-            return matchingInvocations.Any() && !IsReturnValueUsed(matchingInvocations);
-        }
+            var localFunctionSyntax = (LocalFunctionStatementSyntaxWrapper)context.Node;
+            var topmostContainingMethod = context.Node.GetTopMostContainingMethod();
 
-        private static bool IsReturnValueUsed(IEnumerable<SyntaxNodeSymbolSemanticModelTuple<InvocationExpressionSyntax, IMethodSymbol>> matchingInvocations) =>
-            matchingInvocations.Any(invocation =>
-                !IsExpressionStatement(invocation.SyntaxNode.Parent) &&
-                !IsActionLambda(invocation.SyntaxNode.Parent, invocation.SemanticModel));
-
-
-        private static bool IsActionLambda(SyntaxNode node, SemanticModel semanticModel)
-        {
-            if (!(node is LambdaExpressionSyntax lambda))
+            if (!(context.SemanticModel.GetDeclaredSymbol(localFunctionSyntax) is IMethodSymbol localFunctionSymbol)
+                || localFunctionSymbol.ReturnsVoid
+                || topmostContainingMethod == null)
             {
-                return false;
+                return;
             }
 
-            return semanticModel.GetSymbolInfo(lambda).Symbol is IMethodSymbol symbol && symbol.ReturnsVoid;
+            var matchingInvocations = GetLocalMatchingInvocations(topmostContainingMethod, localFunctionSymbol, context.SemanticModel);
+            // Method invocation is noncompliant when there is at least 1 inovcation of the method, and no invocation is using the return value. The case of 0 invocation is handled by S1144.
+            if (matchingInvocations.Any() && !matchingInvocations.Any(x => IsReturnValueUsed(x)))
```

Comment:
It's debatable, but in cases like this I would better convert matchingInvocations to list

Reply:
The same way you do this on line 74